src/libpgmoneta/walfile/relpath.c

```diff
@@ -37,129 +37,170 @@
-                                              dbNode, relNode,
-                                              forkNames[forkNumber]);
-         }
-         else
-         {
-            path = pgmoneta_format_and_append(path, "pg_tblspc/%u/%s/%u/%u",
-                                              spcNode, version_directory,
-                                              dbNode, relNode);
-         }
+      if (backendId == INVALID_BACKEND_ID)
+      {
+         path = pgmoneta_format_and_append(path,
+                                           forkNumber != MAIN_FORKNUM ?
+                                           "pg_tblspc/%u/%s/%u/%u_%s" : "pg_tblspc/%u/%s/%u/%u",
+                                           spcNode, version_directory, dbNode, relNode,
+                                           forkNumber != MAIN_FORKNUM ? FORK_NAMES[forkNumber] : NULL);
       }
       else
       {
-         if (forkNumber != MAIN_FORKNUM)
-         {
-            path = pgmoneta_format_and_append(path, "pg_tblspc/%u/%s/%u/t%d_%u_%s",
-                                              spcNode, version_directory,
-                                              dbNode, backendId, relNode,
-                                              forkNames[forkNumber]);
-         }
-         else
-         {
-            path = pgmoneta_format_and_append(path, "pg_tblspc/%u/%s/%u/t%d_%u",
-                                              spcNode, version_directory,
-                                              dbNode, backendId, relNode);
-         }
+         path = pgmoneta_format_and_append(path,
+                                           forkNumber != MAIN_FORKNUM ?
+                                           "pg_tblspc/%u/%s/%u/t%d_%u_%s" : "pg_tblspc/%u/%s/%u/t%d_%u",
+                                           spcNode, version_directory, dbNode, backendId, relNode,
+                                           forkNumber != MAIN_FORKNUM ? FORK_NAMES[forkNumber] : NULL);
       }
       free(version_directory);
    }
+
+   if (!path)
+   {
+      goto error;
+   }
    return path;
+
+error:
+   if (path)
+   {
+      free(path);
+   }
+   return NULL;
 }
 
 char*
 pgmoneta_wal_get_tablespace_version_directory(void)
 {
-   char* result = (char*)malloc(50);
-   result = pgmoneta_format_and_append(result, "PG_%d_%s", server_config->version, pgmoneta_wal_get_catalog_version_number());
+   char* result = NULL;
+   char* catalog_version = NULL;
+
+   if (!server_config)
+   {
+      goto error;
+   }
+
+   result = (char*)malloc(MAX_VERSION_DIR_SIZE);
+   if (!result)
+   {
+      goto error;
+   }
+
+   catalog_version = pgmoneta_wal_get_catalog_version_number();
+   if (!catalog_version)
+   {
+      goto error;
+   }
+
+   if (!pgmoneta_format_and_append(result, "PG_%d_%s",
+                                   server_config->version, catalog_version))
+   {
+      goto error;
+   }
+
    return result;
+
+error:
+   if (result)
+   {
+      free(result);
```

Comment:
if is not needed, free(NULL)
`If ptr is a null pointer, the function does nothing.`